src/libopensc/card-eoi.c

```diff
@@ -0,0 +1,544 @@
+
+	if (privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	privdata = sc_mem_secure_alloc(sizeof(struct eoi_privdata));
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+	/* sc_mem_secure_alloc()-ed memory may not be zeroized */
+	memset(privdata, 0, sizeof(struct eoi_privdata));
+	card->drv_data = privdata;
+
+	sprintf(privdata->version, "%X%02X.%02X%02X", version[0], version[1], version[2], version[3]);
+	sc_log(card->ctx, "App version: %s", privdata->version);
+
+	memset(&card->sm_ctx, 0, sizeof card->sm_ctx);
+	card->sm_ctx.ops.open = eoi_sm_open;
+
+	card->max_send_size = SC_MAX_APDU_RESP_SIZE;
+	card->max_recv_size = SC_MAX_APDU_RESP_SIZE;
+
+	for (i = 0; i < sizeof eoi_curves / sizeof * eoi_curves; ++i) {
+		int r = _sc_card_add_ec_alg(card, eoi_curves[i].len, SC_ALGORITHM_ECDSA_RAW | SC_ALGORITHM_ECDSA_HASH_NONE, 0, &eoi_curves[i].oid);
+		LOG_TEST_RET(card->ctx, r, "Add EC alg failed");
+	}
+
+	if (getenv("EOI_CAN"))
+		strncpy(privdata->can, getenv("EOI_CAN"), sizeof(privdata->can) - 1);
+	for (i = 0; card->ctx->conf_blocks[i]; i++) {
+		found_blocks = scconf_find_blocks(card->ctx->conf, card->ctx->conf_blocks[i],
+					"card_driver", "eoi");
+		if (!found_blocks)
+			continue;
+
+		for (j = 0, block = found_blocks[j]; block; j++, block = found_blocks[j]) {
+			if (!privdata->can[0]) {
+				const char *can = scconf_get_str(block, "can", NULL);
+				if (can)
+					strncpy(privdata->can, can, sizeof(privdata->can) - 1);
+			}
+		}
+	}
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int eoi_finish(sc_card_t* card)
+{
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (privdata) {
+		sc_mem_clear(privdata, sizeof(struct eoi_privdata));
+		sc_mem_secure_free(privdata, sizeof(struct eoi_privdata));
+	}
+
+	card->drv_data = NULL;
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int eoi_select_file(sc_card_t *card, const sc_path_t *in_path, sc_file_t **file_out)
+{
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+	int i;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	for (i = 0; i < MAX_OBJECTS && privdata->pin_paths[i]; i++) {
+		if (privdata->pin_paths[i] && sc_compare_path(privdata->pin_paths[i], in_path)) {
+			LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+		}
+	}
+
+	LOG_FUNC_RETURN(card->ctx, sc_get_iso7816_driver()->ops->select_file(card, in_path, file_out));
+}
+
+static int eoi_logout(struct sc_card *card)
+{
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+	struct sc_apdu apdu;
+	u8 buf[256];
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	iso_sm_close(card);
+	card->sm_ctx.sm_mode = SM_MODE_NONE;
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0xCA, 0x00, 0x00);
+	apdu.cla = 0xFF;
+	apdu.resp = buf;
+	apdu.resplen = 256;
+	apdu.lc = 0;
+	apdu.le = 256;
```

Comment:
APDUs with CLA=FF are typically interpreted by the reader. Could it be that you actually want to call DestroyPACEChannel from BSI TR-03119?
![grafik](https://user-images.githubusercontent.com/1266577/217380783-d201507e-d136-4f15-84b2-8dc56bff8f42.png)
If so, it may be more useful to explicitly add a callback to reader->ops for destroy_pace and extend reader-tr03119.c to implement this callback with pseudo apdus accordingly.

Reply:
but please leave the APDU still present in the code

Reply:
but please leave the APDU still present in the code so that it's still executed

Reply:
Done.

Reply:
ah right, now I remember: This command actually gets the UID if the card is contactless and the *reader* supports it.  I already implemented this in `initialize_uid()` in reader-pcsc.c. This command is defined in the PC/SC standard and doesn't actually reset the card.

In OpenSC we use this to detect whether we are connected via the contact-less interface. I would assume that the official client also uses this for some kind of detection rather than some kind of reset (hence, you could think about removing this). If you want to keep this command in your code for compatibility, I think it would be good to respect the `SC_READER_ENABLE_ESCAPE` in the `reader->flags`.

Reply:
The UID comes from the ISO 14443-3 standard and is some kind of NFC identification used to differentiate multiple cards in the field. `08` says that your UID is generated at random on every reset of the contactless field. the other three bytes contain the randomized ID.

Reply:
I'v added a check for `SC_READER_ENABLE_ESCAPE`, however it isn't set anywhere in a pcsc reader driver?

Reply:
oh... I'll check in seperately